backend/ohq/permissions.py

```diff
@@ -503,3 +503,107 @@ def has_permission(self, request, view):
+            return membership.is_ta
+
+        # Head TAs+ can make changes
+        if view.action in ["retrieve", "create", "destroy", "update", "partial_update"]:
+            return membership.is_leadership
+
+class DocumentPermission(permissions.BasePermission):
+    def has_object_permission(self, request, view, obj):
+        
+        membership = Membership.objects.get(course=view.kwargs["course_pk"], user=request.user)
+
+        if view.action == "retrieve":
+            return membership.is_ta
+
+        if view.action in ["retrive", "destroy", "partial_update", "update"]:
+            return membership.is_leadership
+        
+        return False
+
+    def has_permission(self, request, view):
+        if not request.user.is_authenticated:
+            return False
+
+        membership = Membership.objects.filter(
+            course=view.kwargs["course_pk"], user=request.user
+        ).first()
+
+        # Non-Students can't do anything
+        if membership is None:
+            return False
+
+        if view.action == "retrieve":
+            return membership.is_ta
+
+        # Head TAs+ can make changes
+        if view.action in ["retrieve", "create", "destroy", "update", "partial_update"]:
+            return membership.is_leadership
+        
+class VectorSearchPermission(permissions.BasePermission):
+    def has_permission(self, request, view):
+        if not request.user.is_authenticated:
+            return False
+
+        membership = Membership.objects.filter(
+            course=view.kwargs["course_pk"], user=request.user
+        ).first()
+
+        # Non-Students can't do anything
+        if membership is None:
+            return False
+
+        if view.action == "retrieve":
+            return membership.is_ta
+
+        # Head TAs+ can make changes
+        if view.action in ["retrieve", "create", "destroy", "update", "partial_update"]:
+            return membership.is_leadership
+    
+class VectorDBPermission(permissions.BasePermission):
+    def has_object_permission(self, request, view, obj):
+
+        membership = Membership.objects.get(course=view.kwargs["course_pk"], user=request.user)
+
+        if view.action == "retrieve":
+            return membership.is_ta
+
+        if view.action in ["retrieve", "destroy", "partial_update", "update"]:
+            return membership.is_leadership
+        
+        return False
+
+    def has_permission(self, request, view):
+        if not request.user.is_authenticated:
+            return False
+
+        membership = Membership.objects.filter(
+            course=view.kwargs["course_pk"], user=request.user
+        ).first()
+
+        # Non-Students can't do anything
+        if membership is None:
+            return False
+
+        if view.action == "retrieve":
+            return membership.is_ta
+
+        # Head TAs+ can make changes
+        if view.action in ["retrieve", "create", "destroy", "update", "partial_update"]:
```

Comment:
Same comment as in the DocumentCreatePermission class! Remove retrieve if retrieve permission is already defined above!